### PR TITLE
Activity regularization

### PR DESCRIPTION
**Pull request by phreeza**
As discussed in #231 

I added a way to add updates to the global cost function from a layer. This has the advantage that we can let theano calculate the gradient, unlike the way it is in the other regularizers right now where we need to supply the gradient. Maybe all regularizers could be handled like this in the future?

I will work on an implementation of the sparse denoising autoencoder tomorrow, there is a trivial demo included already.

### CODE REVIEW

File: keras/models.py [hunk 1]
```diff
@@ -321,6 +325,7 @@ def __init__(self):
         self.params = [] # learnable
         self.regularizers = [] # same size as params
         self.constraints = [] # same size as params
+        self.cost_updates = [] # NOT the same size as params
```
Comment (fchollet) [inline keras/models.py]: Consistency?

File: keras/models.py [hunk 2]
```diff
@@ -87,7 +87,11 @@ def compile(self, optimizer, loss, class_mode="categorical", theano_mode=None):
             raise Exception("Invalid class mode:" + str(class_mode))
         self.class_mode = class_mode
 
-        updates = self.optimizer.get_updates(self.params, self.regularizers, self.constraints, train_loss)
+        if hasattr(self, 'cost_updates'):
+            for u in self.cost_updates:
+                train_loss += u()
```
Comment (fchollet) [inline keras/models.py]: Definitely a nice way to implement regularization. We'll have to think about a unified system for integrating regularization as part of the loss.

File: keras/regularizers.py [hunk 3]
```diff
@@ -23,4 +23,23 @@ def l1l2wrap(g, p):
...
-    return g
\ No newline at end of file
+    return g
+
+
+def activity_l1(l=.01):
+    # activity dependent l1 norm
+    def l1wrap(layer):
+        # needs to be wrapped twice because input is not present during instantiation
+        def l1wrap_wrap():
```
Comment (fchollet) [inline keras/regularizers.py]: But who will wrap the wrappers? `l1wrap_wrap`, apparently.

File: keras/models.py (hunk 2 above)
Comment (phreeza) [inline keras/models.py]: I think the existing regularisations could be turned into this style quite easily, should we do that?

File: keras/models.py (hunk 1 above)
Comment (phreeza) [inline keras/models.py]: You mean the different lengths? I don't think it can be helped.

File: keras/models.py (hunk 2 above)
Comment (fchollet) [inline keras/models.py]: Yes, that would be a good idea. We would need to monitor accuracy and speed along the way. Incorporating regularization in the loss should result in higher accuracy but *might* cause performance issues.

File: keras/regularizers.py (hunk 3 above)
Comment (fchollet) [inline keras/regularizers.py]: Would there be a better way to handle this?

File: keras/models.py (hunk 1 above)
Comment (fchollet) [inline keras/models.py]: Yes, it's a presentation/comments issue. The comment "same size as params" means there is a 1:1 mapping between the arrays. The comment "NOT the same size as params" is not informative as to what the array contains.

File: keras/regularizers.py (hunk 3 above)
Comment (phreeza) [inline keras/regularizers.py]: Probably the way to go would be to make regularizers a class. We could have a class that includes both the old (gradient updating) and new (cost updating) functions. They would then implement either get_cost_update or get_gradient_update.

How do you feel about that? I am a bit busy right now but I could get to it around the middle of the week.

**Comment by phreeza** [comment]
I have refactored the regularizer system to be object oriented, I think it is more elegant and futureproof this way. I managed to keep it backwards compatible, so these changes should not break any existing code.

It leaves the option open to transition to purely cost-based regularization in the future, if desired.

I also added proper unit tests. These use the old style of creating regularizers.

File: keras/regularizers.py [hunk 4]
```diff
@@ -1,26 +1,117 @@
...
+        self.modifies_cost = False
+
+    def update_gradient(self, gradient, params):
+        raise NotImplementedError
+
+    def update_cost(self, cost):
+        raise NotImplementedError
+
+
+class RegularizerWeightsL1(Regularizer):
```
Comment (fchollet) [inline keras/regularizers.py]: What about simply `L1`? It lives inside `regularizers` and inherits from `Regularizer`, so it's pretty clear it's a regularizer.

File: keras/layers/containers.py [hunk 5]
```diff
@@ -20,6 +20,8 @@ def __init__(self, layers=[]):
         self.params = []
         self.regularizers = []
         self.constraints = []
+        self.cost_updates = []
```
Comment (fchollet) [inline keras/layers/containers.py]: The general term used throughout Keras is "loss" (as well as "objective" for the objective functions); I think we should keep it that way and replace "cost" with "loss" in the variables.

File: keras/models.py [hunk 6]
```diff
@@ -89,7 +89,12 @@ def compile(self, optimizer, loss, class_mode="categorical", theano_mode=None):
             raise Exception("Invalid class mode:" + str(class_mode))
         self.class_mode = class_mode
 
-        updates = self.optimizer.get_updates(self.params, self.regularizers, self.constraints, train_loss)
+        if hasattr(self, 'cost_updates'):
+            for u in self.cost_updates:
+                if u.modifies_cost:
+                    train_loss = u(train_loss)
```
Comment (fchollet) [inline keras/models.py]: You can get rid of this parameter (and its check in `compile`) by using:
`train_loss += u.loss() `

(which would return 0 if the update does not affect the loss).

File: keras/regularizers.py [hunk 7]
```diff
@@ -1,26 +1,117 @@
...
+        return cost + self.l * T.sum(T.mean(self.layer.get_output(True) ** 2, axis=0))
+
+    def __call__(self, cost):
+        return self.update_cost(cost)
+
+#old style variables for backwards compatibility
+l1 = RegularizerWeightsL1
+l2 = RegularizerWeightsL2
+l1l2 = RegularizerWeightsL1L2
+identity = RegularizerIdentity()
```
Comment (fchollet) [inline keras/regularizers.py]: That's cool. For consistency with activity we should have something like

`l2 = weights_l2 = WeightsL2`

**Comment by fchollet** [comment]
I like the move to a `Regularizer` class. I left a few comments!

**Comment by phreeza** [comment]
Hi François, thanks for your feedback.

I have fixed the other suggestions, but I am a bit unsure what you mean about removing the flags. The advantage of `train_loss = u(train_loss)` instead of  `train_loss += u.loss()` is that it allows more complex interactions such as setting the gradient to zero. I am sure there are other scenarios where something more than additive cost functions might be useful?

But of course those two flags `modifies_cost` and `modifies_gradient` are needed to achieve that. I feel like they can also serve as an additional sanity check. I was also thinking the flags could make it possible to store all cost updates in a single list at some point in the future.

Let me know what you think, Thomas

**Comment by fchollet** [comment]
> The advantage of train_loss = u(train_loss) instead of train_loss += u.loss() is that it allows more complex interactions such as setting the gradient to zero. I am sure there are other scenarios where something more than additive cost functions might be useful?

Right, in that case a better syntax would be `train_loss = u.update_loss(train_loss)`. 

What I mean by getting of the flags is:
```python
# instead of:
if u.modifies_loss:
   loss = u(loss)

# use
loss = u.update_loss(loss) # can also leave the loss unchanged...

# instead of:
if u.modifies_gradient:
   g, p = u(g, p)

# use
g, p = u.update_gradient(g, p) # can also leave g, p unchanged...
```

If does the same, but without any control flow necessary.

**Comment by phreeza** [comment]
OK, now I understood and fixed that as well. The only downside I see with this is that it breaks any custom regularisers that people may have written earlier. But I agree this makes the structure clearer when using them.

File: keras/regularizers.py [hunk 8]
```diff
@@ -1,26 +1,80 @@
...
+
+    def update_loss(self, loss):
+        return loss + self.l * T.sum(T.mean(self.layer.get_output(True) ** 2, axis=0))
+
+
+#old style variables for backwards compatibility
+l1 = weights_l1 = WeightsL1
+l2 = weights_l2 = WeightsL2
+l1l2 = weights_l1l2 = WeightsL1L2
+identity = Identity()
```
Comment (fchollet) [inline keras/regularizers.py]: Should be the class `Identity`, not an instance thereof...?

File: keras/regularizers.py [hunk 9]
```diff
@@ -1,26 +1,80 @@
...
+        self.l1 = l1
+        self.l2 = l2
+
+    def update_gradient(self, gradient, params):
+        gradient += params * self.l2
+        gradient += T.sgn(params) * self.l1
+        return gradient
+
+
+class Identity(Regularizer):
```
Comment (fchollet) [inline keras/regularizers.py]: Could definitely also implement `update_loss` (return loss unchanged).

File: keras/models.py [hunk 10]
```diff
@@ -370,4 +379,33 @@ def load_weights(self, filepath):
             g = f['layer_{}'.format(k)]
             weights = [g['param_{}'.format(p)] for p in range(g.attrs['nb_params'])]
             self.layers[k].set_weights(weights)
-        f.close()
\ No newline at end of file
+        f.close()
+
+
+class Autoencoder(Sequential):
```
Comment (fchollet) [inline keras/models.py]: What is this model doing in this PR, given that autoencoders are being implemented with a layer in `core.py`...?

File: keras/models.py [hunk 11]
```diff
@@ -89,7 +93,11 @@ def compile(self, optimizer, loss, class_mode="categorical", theano_mode=None):
             raise Exception("Invalid class mode:" + str(class_mode))
         self.class_mode = class_mode
 
-        updates = self.optimizer.get_updates(self.params, self.regularizers, self.constraints, train_loss)
+        if hasattr(self, 'cost_updates'):
```
Comment (fchollet) [inline keras/models.py]: Should be `loss_updates`

File: keras/models.py (hunk 10 above)
Comment (phreeza) [inline keras/models.py]: Ugh, something must have gone wrong when merging recent changes from master...

File: keras/regularizers.py (hunk 8 above)
Comment (phreeza) [inline keras/regularizers.py]: This is for backwards compatibility, because the old identity was not a wrapped function, while the others were. I am fine with breaking that if you are, would be more consistent.

Comment (fchollet) [inline keras/regularizers.py]: Let's definitely prioritize consistency and code quality over backwards compatibility. Keras is in alpha and subject to occasional breaking changes.

File: keras/layers/core.py [hunk 12]
```diff
@@ -80,8 +80,6 @@ def __init__(self, models, mode='sum'):
         ''' Merge the output of a list of models into a single tensor.
             mode: {'sum', 'concat'}
         '''
-        if len(models) < 2:
-            raise Exception("Please specify two or more input models to merge")
```
Comment (fchollet) [inline keras/layers/core.py]: Why remove this check?

File: keras/regularizers.py (hunk 9 above)
Comment (phreeza) [inline keras/regularizers.py]: Indeed, actually could do that in the base class.

But do we not maybe want to keep it like this so that an error gets raised in the event that one of these regularizers gets put into a place where it does not belong? (like putting an activity regularizer in the slot for a weight regularizer?)

File: keras/layers/core.py (hunk 12 above)
Comment (phreeza) [inline keras/layers/core.py]: Sorry, I accidentally included my old autoencoder design in this PR, this was part of it. No idea how that happened, should be all fixed now.

File: keras/regularizers.py (hunk 9 above)
Comment (fchollet) [inline keras/regularizers.py]: The idea would be that Identity could serve as both the identity weight regularizer and the identity activity regularizer.

File: tests/auto/test_regularizers.py [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+nb_classes = 10
+batch_size = 128
+nb_epoch = 5
+weighted_class = 9
+standard_weight = 1
+high_weight = 5
+max_train_samples = 5000
+max_test_samples = 1000
+
+np.random.seed(1337) # for reproducibility
```
Comment (fchollet) [inline tests/auto/test_regularizers.py]: This line actually belong before any imports from layers.core.

**Comment by fchollet** [comment]
Looks good! We'll merge this. Apart from the small details we mentioned, we should definitely update the doc to introduce the changes.

File: keras/models.py [hunk 14]
```diff
@@ -7,7 +7,9 @@
 
 from . import optimizers
 from . import objectives
-from . import callbacks as cbks
```
Comment (fchollet) [inline keras/models.py]: This line is required...

**Comment by fchollet** [comment]
I remove the loss_update system, which did not make sense to me in its current form. Which should reintroduce activity regularization as part of a unified way to handle regularization via the loss function. I'll look at it later today.

**Comment by fchollet** [comment]
Here's what I propose: regularizers get instantiated with the quantity that they regularize (weight parameter or layer output). A list of such regularizers (of arbitrary length) is then used to update the loss function of a model.
```python
class Regularizer(object):
    def __call__(self, loss):
        return loss


class WeightRegularizer(Regularizer):
    def __init__(self, p, l1=0., l2=0.):
        self.p = p
        self.l1 = l1
        self.l2 = l2

    def __call__(self, loss):
        loss += T.sgn(self.p) * self.l1
        loss += self.p * self.l2
        return loss


class ActivityRegularizer(Regularizer):
    def __init__(self, output, l1=0., l2=0.):
        self.l1 = l1
        self.l2 = l2
        self.output = output

    def set_layer(self, layer):
        self.layer = layer

    def __call__(self, loss):
        loss += self.l1 * T.sum(T.mean(abs(self.output), axis=0))
        loss += self.l2 * T.sum(T.mean(self.output ** 2, axis=0))
        return loss
```

**Comment by fchollet** [comment]
Done, all regularization is now happening via the loss function. Activity regularization is back as well, both through the dedicated ActivityRegularization layer, and through arguments allowed in certain layers (Dense layers, Convolution layers).

**Comment by kenterao** [comment]
I agree that making regularizer classes makes things cleaner.  

What do you think about adding ```get_config() ``` to the ```Regularizer``` class so a layer can be fully reconstructed from it's config.  

```python
class WeightRegularizer(Regularizer):
    def __init__(self, l1=0., l2=0.):
        self.l1 = l1
        self.l2 = l2

    def get_config(self):
        return {'name': self.__class__.__name__,
            'l1': self.l1, 'l2': self.l2}

class Dense(Layer):
    def get_config(self):
        return {"name": self.__class__.__name__,
            "W_regularizer": self.W_regularizer.get_config(),
            "b_regularizer": self.b_regularizer.get_config(),
            ...
        }
```

**Comment by phreeza** [comment]
I like what you did with the regularizers, @fchollet , I guess I was too worried about compatibility, keeping the gradient-updating regularizers :-)

@kenterao's suggestion sounds good to me.
